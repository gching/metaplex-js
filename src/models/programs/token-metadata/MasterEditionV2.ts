import { PublicKey } from '@solana/web3.js';
import BN from 'bn.js';
import { deserializeUnchecked } from 'borsh';
import { BorshSchema } from '../../utils/types';
import { MetadataKey } from './MetadataKey';
import { AccountWithPublicKey } from '../../../utils/privateTypes';

export class MasterEditionV2 {
  static SCHEMA: BorshSchema = new Map([
    [
      MasterEditionV2,
      {
        kind: 'struct',
        fields: [
          ['key', 'u8'],
          ['supply', 'u64'],
          ['maxSupply', { kind: 'option', type: 'u64' }],
        ],
      },
    ],
  ]);

  key = MetadataKey.MasterEditionV2;
  accountPublicKey?: PublicKey;
  supply: BN;
  maxSupply?: BN;

  constructor(args: Omit<MasterEditionV2, 'key'>) {
    this.accountPublicKey = args.accountPublicKey;
    this.supply = args.supply;
    this.maxSupply = args.maxSupply;
  }

  static decode(accountWithPublicKey: AccountWithPublicKey): MasterEditionV2 {
    const masterEdition = deserializeUnchecked(
      MasterEditionV2.SCHEMA,
      MasterEditionV2,
      accountWithPublicKey.account.data,
    ) as MasterEditionV2;

    return new MasterEditionV2({
      ...masterEdition,
      accountPublicKey: accountWithPublicKey.pubkey,
    });
  }
}
