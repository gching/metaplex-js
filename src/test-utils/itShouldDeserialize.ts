import { PublicKey } from '@solana/web3.js';
import { serialize } from 'borsh';
import { BorshSchema } from '../models/utils/types';

export function itShouldDeserialize<E>(
  cls: {
    SCHEMA: BorshSchema;
    // eslint-disable-next-line @typescript-eslint/ban-types
    decode: Function;
  },
  received: E,
  accountPubKey = PublicKey.default,
): void {
  const data = Buffer.from(serialize(cls.SCHEMA, received));

  const deserialized = cls.decode({
    pubkey: accountPubKey,
    account: {
      data,
    },
  }) as E;

  expect(deserialized).toDeepEqual(received);
}
