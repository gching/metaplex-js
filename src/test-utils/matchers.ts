import { isEqualWith } from 'lodash';
import BN from 'bn.js';

declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace jest {
    interface Matchers<R> {
      toDeepEqual: <E = any>(expected: E) => R;
    }
  }
}

expect.extend({
  toDeepEqual<E = any>(received: E, expected: E) {
    const pass = isEqualWith(received, expected, (objValue, otherValue) => {
      if (BN.isBN(objValue) && BN.isBN(otherValue)) {
        return objValue.eq(otherValue);
      }
      return undefined;
    });

    if (pass) {
      return {
        // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
        message: () => `expected ${received} not to deep equal ${expected}`,
        pass: true,
      };
    }

    return {
      // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
      message: () => `expected ${received} to deep equal ${expected}`,
      pass: false,
    };
  },
});

export default undefined;
