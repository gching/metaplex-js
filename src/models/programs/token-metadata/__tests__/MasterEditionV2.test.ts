import { PublicKey } from '@solana/web3.js';
import BN from 'bn.js';
import { itShouldDeserialize } from '../../../../test-utils/itShouldDeserialize';
import { MasterEditionV2 } from '../MasterEditionV2';

describe('MasterEditionV2', () => {
  it('should instantiate with the correct values', () => {
    const inst = new MasterEditionV2({
      accountPublicKey: PublicKey.default,
      supply: new BN(0),
    });

    expect(inst).toMatchSnapshot();
  });

  it('should deserialize with the correct values', () => {
    const inst = new MasterEditionV2({
      accountPublicKey: PublicKey.default,
      supply: new BN(0),
    });

    itShouldDeserialize(MasterEditionV2, inst);
  });
});
