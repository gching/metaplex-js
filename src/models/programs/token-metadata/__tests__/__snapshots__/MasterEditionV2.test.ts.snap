// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MasterEditionV2 should instantiate with the correct values 1`] = `
MasterEditionV2 {
  "accountPublicKey": PublicKey {
    "_bn": "00",
  },
  "key": 6,
  "maxSupply": undefined,
  "supply": "00",
}
`;
